File: vaku2/path_update.go
package vaku2

import (
	"errors"
)

var (
	ErrPathUpdate = errors.New("path update")
)

// PathUpdate updates a path with data. Existing data is merged with new data. Precedence is given
// to new data.
func (c *Client) PathUpdate(p string, d map[string]interface{}) error {
	if d == nil {
		return newWrapErr(p, ErrPathUpdate, ErrNilData)
	}

	read, err := c.PathRead(p)
	if err != nil {
		return newWrapErr(p, ErrPathUpdate, err)
	}
	if read == nil {
		read = make(map[string]interface{}, len(d))
	}

	for k, v := range d {
		read[k] = v
	}

	err = c.PathWrite(p, read)
	if err != nil {
		return newWrapErr(p, ErrPathUpdate, err)
	}

	return nil
}

File: src/api/path_update.rs
//! `PathUpdate` — read-merge-write convenience.

use crate::api::client::Client;
use crate::api::error::{Error, ErrorKind};
use serde_json::{Map, Value};

impl Client {
    /// Update a path by reading existing data, merging `data` into it (new
    /// keys win), and writing back.
    pub async fn path_update(
        &self,
        p: &str,
        data: Option<Map<String, Value>>,
    ) -> Result<(), Error> {
        let Some(data) = data else {
            return Err(Error::wrap(
                p,
                ErrorKind::PathUpdate,
                Some(Box::new(Error::wrap("", ErrorKind::NilData, None))),
            ));
        };

        let mut read = match self.path_read(p).await {
            Ok(Some(m)) => m,
            Ok(None) => Map::new(),
            Err(e) => return Err(Error::wrap(p, ErrorKind::PathUpdate, Some(Box::new(e)))),
        };

        for (k, v) in data {
            read.insert(k, v);
        }

        self.path_write(p, Some(read))
            .await
            .map_err(|e| Error::wrap(p, ErrorKind::PathUpdate, Some(Box::new(e))))
    }
}
